zio-http/src/main/scala/zio/http/Request.scala

```diff
@@ -1,199 +1,55 @@
+   * Updates the headers using the provided function
    */
-  def version: Version
-
-  private[zio] trait UnsafeAPI {
-
-    /**
-     * Accesses the channel's context for more low level control
-     */
-    def context(implicit unsafe: Unsafe): Ctx
-
-    /**
-     * Gets the HttpRequest
-     */
-    def encode(implicit unsafe: Unsafe): HttpRequest
-  }
-
-  private[zio] val unsafe: UnsafeAPI
+  override def updateHeaders(update: Headers => Headers): Request = self.copy(headers = update(self.headers))
 }
 
 object Request {
 
   /**
-   * Constructor for Request
+   * Constructor for Request that provides sane defaults.
    */
-  def apply(
+  def make(
     version: Version = Version.`HTTP/1.1`,
     method: Method = Method.GET,
     url: URL = URL.root,
     headers: Headers = Headers.empty,
     body: Body = Body.empty,
-  ): Request = {
-    val m = method
-    val u = url
-    val h = headers
-    val d = body
-    val v = version
-
-    new Request {
-      override final val method: Method   = m
-      override final val url: URL         = u
-      override final val headers: Headers = h
-      override final val version: Version = v
-      override final val body: Body       = d
-
-      override final val unsafe: UnsafeAPI = new UnsafeAPI {
-        override final def context(implicit unsafe: Unsafe): Ctx        = throw new IOException(
-          "Request does not have a context",
-        )
-        override final def encode(implicit unsafe: Unsafe): HttpRequest = {
-          val jVersion = v.toJava
-          val path     = url.relative.encode
-          new DefaultFullHttpRequest(jVersion, method.toJava, path)
-        }
-      }
-    }
-  }
-
-  private[zio] def fromFullHttpRequest(jReq: FullHttpRequest)(implicit ctx: Ctx): Request = {
-    val protocolVersion = Version.unsafe.fromJava(jReq.protocolVersion())(Unsafe.unsafe)
-
-    new Request {
-      override final def method: Method    = Method.fromHttpMethod(jReq.method())
-      override final def url: URL          = URL.fromString(jReq.uri()).getOrElse(URL.empty)
-      override final def headers: Headers  = Headers.make(jReq.headers())
-      override final def body: Body        = Body.fromByteBuf(jReq.content())
-      override final val version: Version  = protocolVersion
-      override final val unsafe: UnsafeAPI = new UnsafeAPI {
-        override final def encode(implicit unsafe: Unsafe): HttpRequest = jReq
-        override final def context(implicit unsafe: Unsafe): Ctx        = ctx
-      }
-    }
-  }
-
-  private[zio] def fromHttpRequest(jReq: HttpRequest)(implicit ctx: Ctx): Request = {
-    val protocolVersion = Version.unsafe.fromJava(jReq.protocolVersion())(Unsafe.unsafe)
-
-    new Request {
-      override final def headers: Headers = Headers.make(jReq.headers())
-      override final def method: Method   = Method.fromHttpMethod(jReq.method())
-      override final def url: URL         = URL.fromString(jReq.uri()).getOrElse(URL.empty)
```

Comment:
Looking at these, I suspect the use of `trait` for `Request` was not buying any performance gains, because every field is created every time. In addition, you succeeded to delete `val unsafe` which has negative performance implications.

Reply:
Deleting unsafe from all the public APIs is one of my end-goals for this effort.